### PR TITLE
Fix auth and implement following backend

### PR DESCRIPTION
**Pull request by alicesf2**
fix auth, implemented following a project in backend

### CODE REVIEW

**Comment by arpanlaha** [comment]
Cat Gif

File: api/public/schema/projectSchema.js [hunk 1]
```diff
@@ -38,6 +38,10 @@ const schema = {
       type: "string",
       required: false
     },
+    followers: {
```
Comment (arpanlaha) [inline api/public/schema/projectSchema.js:41]: shouldn't this be an array of strings?

File: auth-server/src/api/register.js [hunk 2]
```diff
@@ -28,6 +25,7 @@ router.post(
       .isLength({ min: 1 })
   ],
   handleAsyncErrors(async function(req, res) {
+    console.log(req.body);
```
Comment (arpanlaha) [inline auth-server/src/api/register.js:28]: leftover log

File: auth-server/src/api/register.js [hunk 3]
```diff
@@ -63,26 +61,29 @@ router.post(
       }
       const question =
         securityQuestionsResponse.securityQuestions[req.body.questionIdx];
-      if (!question || !req.body.answer) {
+      console.log(question);
```
Comment (arpanlaha) [inline auth-server/src/api/register.js:64]: another leftover

File: auth-server/src/api/register.js [hunk 4]
```diff
@@ -63,26 +61,29 @@ router.post(
...
-      );
-    }
+    // const requiredAuthFrom = await getRolesForUser(req.body.role);
+    // if (requiredAuthFrom != null) {
+    //   return sendResponse(
+    //     res,
+    //     400,
+    //     "User needs a higher permission level for that role"
+    //   );
+    // }
```
Comment (arpanlaha) [inline auth-server/src/api/register.js:86]: can remove if we don't need it

File: auth-server/src/utils/jwtHelpers.js [hunk 5]
```diff
@@ -1,82 +1,68 @@
...
-      let { userId, hashedPassword } = jwt.verify(
-        token,
-        String(process.env.AUTH_SECRET[1])
-      );
-      return String(userId) === String(id) && hashedPassword == password;
-    }
-    return false;
+    return err;
   }
+  // } catch (err) {
```
Comment (arpanlaha) [inline auth-server/src/utils/jwtHelpers.js:35]: more leftover comments

File: auth-server/src/utils/jwtHelpers.js [hunk 6]
```diff
@@ -1,82 +1,68 @@
...
-          token,
-          String(process.env.AUTH_SECRET[1])
-        );
-        return userId;
-      } catch (err) {
-        return null;
-      }
-    }
-    return null;
+    // if (process.env.AUTH_SECRET.length > 1) {
```
Comment (arpanlaha) [inline auth-server/src/utils/jwtHelpers.js:53]: same as above

File: api/public/schema/projectSchema.js (hunk 1 above)
Comment (alicesf2) [reply api/public/schema/projectSchema.js:41]: wait that's weird it still works with my following backend stuff

File: auth-server/src/utils/jwtHelpers.js [hunk 7]
```diff
@@ -1,82 +1,46 @@
...
+  });
 }
 
 // Return true if the JWT is valid and matches the parameters
-function verifyAuthJWT(token, id, password) {
+function verifyAuthJWT(token, userEmail) {
   try {
-    let { userId, hashedPassword } = jwt.verify(token, process.env.AUTH_SECRET);
-    if (String(userId) === String(id) && hashedPassword == password) {
+    let { email } = jwt.verify(token, process.env.AUTH_SECRET);
```
Comment (yousefa00) [inline auth-server/src/utils/jwtHelpers.js:15]: What happens if the user passes a token that does not contain an email field. Does jwt.verify return a proper error that will be handled by line 16?

File: auth-server/src/utils/jwtHelpers.js [hunk 8]
```diff
@@ -1,82 +1,46 @@
...
+    if (String(userEmail) === String(email)) {
       return true;
     }
   } catch (err) {
     console.log("Token was updated");
```
Comment (yousefa00) [inline auth-server/src/utils/jwtHelpers.js:20]: Why does this log "Token was updated"? Maybe invalid token would be more explicit.

File: auth-server/src/utils/jwtHelpers.js [hunk 9]
```diff
@@ -1,82 +1,46 @@
...
-function shouldUpdateJWT(token, id, password) {
+function shouldUpdateJWT(token, email, role) {
   try {
-    let { userId, hashedPassword } = jwt.verify(token, process.env.AUTH_SECRET);
-    if (String(userId) === String(id) && hashedPassword == password) {
+    let { userEmail, userRole } = jwt.verify(token, process.env.AUTH_SECRET);
+    if (String(userEmail) === String(email) && userRole === role) {
       return false;
     }
-    return false;
```
Comment (yousefa00) [inline auth-server/src/utils/jwtHelpers.js:49]: I think this should be `return false;`. We want to update the users token if the token they are now using is no longer valid by prompting them to go to the login page again or something.

File: auth-server/src/utils/jwtHelpers.js (hunk 7 above)
Comment (alicesf2) [reply auth-server/src/utils/jwtHelpers.js:15]: hmm i'm not too sure about that, i can look into this later!

Comment (alicesf2) [reply auth-server/src/utils/jwtHelpers.js:15]: hm ok i looked at the code for jwt.verity(), i didn't see anything that would return the proper error so i can just check manually

**Pull request merged**